tests/suites/test_suite_psa_crypto_util.function

```diff
@@ -58,20 +58,35 @@ exit:
 /* BEGIN_CASE depends_on:MBEDTLS_PSA_UTIL_HAVE_ECDSA */
 void ecdsa_der_to_raw(int key_bits, data_t *input, data_t *exp_result, int exp_ret)
 {
-    unsigned char *tmp_buf = NULL;
-    size_t tmp_buf_len = exp_result->len;
+    unsigned char *in_buf = NULL;
+    size_t in_buf_len;
+    unsigned char *out_buf = NULL;
+    size_t out_buf_len = exp_result->len;
     size_t ret_len;
 
-    TEST_CALLOC(tmp_buf, tmp_buf_len);
+    TEST_CALLOC(out_buf, out_buf_len);
+
+    /* Verify that parsing of truncated input always fails. */
```

Comment:
I'm not sure I see the point. Considering the input is enclosed in a SEQUENCE, I think all truncated inputs are going to fail at the same point: the initial call to `mbedtls_asn1_get_tag(..., SEQUENCE)`. So, this loop does not provide more coverage than just testing with a buffer of size `input->len - 1`.

In fact, I think we could achieve the same coverage with the function as it was before and just one targeted test case with the input truncated by one byte.

Cc @gilles-peskine-arm as you suggested this. Am I missing something?

Reply:
Test coverage is a balance between several things: coverage of possible scenarios (for which measures such as line coverage are just an approximation), effort to write tests, amount of tests (causing CPU load, disk/bandwidth for logs, time spent by humans sifting through logs, …), etc. The value of having more coverage is greater when the risk and impact of a bug are greater.

For a parser, a buffer overread is a large-impact risk, so it's good to have high coverage of scenarios that could lead to a buffer overread. From a black box perspective, testing all truncations gives fairly high coverage of potential overread scenarios. Doing it systematically has a very low effort: it's at least as much effort to have separate test cases for truncated-by-one (less code, but more data and more worry about coverage).

From a gray box perspective, we know that there's no difference between truncations to a prefix that has type+length+partial data. So we could just test with size-1 and with sizes up to type+length. But that would require more effort to design, write and review the test cases.

So this approach is not something I insist on, but I think it's a cheap way of getting good coverage. I do note that you only proposed testing size-1, but that would miss very small sizes where the type+length parsing could lead to an overread.

Ideally we'd also test a SEQUENCE with truncated content, but there the risk is less of having a buffer overread as opposed to merely wrong parsing, and this would require more effort. In the ASN.1 test suites, we get this kind of testing through the unit tests for the types of the sequence elements, with the gray box knowledge of how their parser is invoked by the SEQUENCE parser.